src/AccountSystem.sol

```diff
@@ -0,0 +1,409 @@
+    /// @param playerTermsSignature The player's signature for the executor terms
+    function play(
+        Action memory action,
+        ExecutorTerms memory terms,
+        bytes memory playerActionSignature,
+        bytes memory playerTermsSignature
+    ) external {
+        if (action.request <= lastRequest) {
+            revert RequestTooLow();
+        }
+        if (terms.rewardTokens.length != terms.basisPoints.length) {
+            revert MismatchedArrayLengths();
+        }
+
+        lastRequest = action.request;
+
+        // Verify signatures
+        bytes32 providedActionHash = actionHash(action);
+        bytes32 providedTermsHash = executorTermsHash(terms);
+
+        if (
+            !SignatureChecker.isValidSignatureNow(
+                player,
+                providedActionHash,
+                playerActionSignature
+            )
+        ) {
+            revert InvalidPlayerActionSignature();
+        }
+        if (
+            !SignatureChecker.isValidSignatureNow(
+                player,
+                providedTermsHash,
+                playerTermsSignature
+            )
+        ) {
+            revert InvalidPlayerTermsSignature();
+        }
+
+        // Execute the game action and handle rewards
+        _play(action, terms);
+    }
+
+    /// @notice Internal function to execute the game action and pay executor rewards
+    /// @param action The game action to execute
+    /// @param terms The executor's compensation terms
+    function _play(Action memory action, ExecutorTerms memory terms) internal {
+        // Record starting balances
+        uint256[] memory startBalances = new uint256[](
+            terms.rewardTokens.length
+        );
+        for (uint256 i = 0; i < terms.rewardTokens.length; i++) {
+            if (terms.rewardTokens[i] == address(0)) {
+                startBalances[i] = address(this).balance;
+            } else {
+                startBalances[i] = IERC20(terms.rewardTokens[i]).balanceOf(
+                    address(this)
+                );
+            }
+        }
+
+        // Execute the game action
+        (bool success, ) = action.game.call{value: action.value}(action.data);
+        if (!success) {
+            revert ActionFailed();
+        }
+
+        // Calculate and pay executor rewards
+        for (uint256 i = 0; i < terms.rewardTokens.length; i++) {
+            uint256 endBalance;
+            if (terms.rewardTokens[i] == address(0)) {
+                endBalance = address(this).balance;
+            } else {
+                endBalance = IERC20(terms.rewardTokens[i]).balanceOf(
+                    address(this)
+                );
+            }
+
+            // Only pay rewards on profit
+            if (endBalance > startBalances[i]) {
+                uint256 profit = endBalance - startBalances[i];
+                uint256 reward = (profit * terms.basisPoints[i]) / 10000;
```

Comment:
Should we have a check to ensure basisPoints are not larger then 10000 or reward is not greater then profit? To prevent potential drain of account by bad actors